design/healcheck-endpoint.md

```diff
@@ -0,0 +1,169 @@
+or integrating with Microprofile Health Subsystem. An additional problem is that Wildfly provides its own
+`HealthReporter` class (see [the code](https://github.com/wildfly/wildfly/blob/master/microprofile/health-smallrye/src/main/java/org/wildfly/extension/microprofile/health/HealthReporter.java))
+that doesn't use groups. The Grouping functionality is essential for this design (explained below).
+
+The biggest downside of providing Keycloak's own implementation is that we'll end up with 2 health endpoints -
+one for Wildfly (`http://localhost:9990/health`) and one for Keycloak (`https://localhost:8443/auth/health`). We
+need to be very clear in the documentation that the latter is the proper one.
+
+Having the above in mind, the design assumes integrating the new Healthcheck API with Keycloak's REST interface.
+The implementation will create additional `/health` endpoint that will parse incoming `timeout` query parameter.
+
+#### Database Healthcheck
+
+One of the most important checks for the Healthcheck API is checking if the database is up. The Check is performed
+in the Passive Mode, meaning it analyzing metrics emitted by JCA subsystem. Currently, the JCA exposes the following metrics:
+- `wildfly_datasources_pool_in_use_count{data_source="KeycloakDS",microprofile_scope="vendor"}`
+- `wildfly_datasources_pool_idle_count{data_source="KeycloakDS",microprofile_scope="vendor"}`
+- `wildfly_datasources_pool_max_wait_time{data_source="KeycloakDS",microprofile_scope="vendor"}`
+- `wildfly_datasources_pool_timed_out{data_source="KeycloakDS",microprofile_scope="vendor"}`
+- `wildfly_datasources_pool_created_count{data_source="KeycloakDS",microprofile_scope="vendor"}`
+The implementation should include some experimentation, but the above 5 metrics should be enough to check the proper
+JCA state. The other metrics are available in [this document](https://docs.google.com/spreadsheets/d/1YGn4wwUupgQsFTmlm-5VXgLYkigTdy_TEnDhqFxNXXo/edit?usp=sharing).
+Another aspect is that JCA subsystem needs to actively probe the data source.
+This can be achieved by configuring `check-valid-connection-sql` and `background-validation`
+(see the [Wildfly Model manual](https://docs.wildfly.org/19/wildscribe/subsystem/datasources/data-source/ExampleDS/index.html)).
+
+#### Healthcheck API for Liveness and Readiness probes
+
+Since Undertow subsystem starts after the database migration, there's no way provide a working REST Health Endpoint
+during the database migration time. The long boot up time may also happen when loading up sessions from the database,
+or a remove Infinispan cluster. In both scenarios, it is necessary to implement a passing Liveness probe, and a failing
+Readiness probe. This will prevent Kubernetes from killing the Pods, but it will also unplug the pods from a Service
+and make them unavailable for the external consumption.
+
+One of the solutions is to use Wildfly's Liveness check available at `http://localhost:9990/health/live`. The check
```

Comment:
Does this mean we'll use WildFly's Liveness check all the time even after KC has started? (That is, at least until WFLY-13931 is resolved.) What if the pod gets into some failing state and needs to be rebooted? WF's endpoint might not be aware of such state. If I understand it correctly, the pod would be just removed from the Service (thanks to the Readiness check) but won't reboot it. Is that enough?

Reply:
Because of https://github.com/keycloak/keycloak-community/pull/55#issuecomment-708356265 there are two situations to analyze here:
- The current Wildfly-based deployment. The answer is yes. If Keycloak gets into such a weird state, it is SRE Team responsibility to look at the SRE Dashboard and verify if Keycloak really hangs. If it does - we need to fix it in our code. The Readiness/Liveness Probe is definitely not a place to do a deep system analysis. It needs to be quick, small and reliable.
- Quarkus-based deployment. In this case Quarkus is tied very closely with Keycloak (it's more a Spring Boot approach rather than JEE Platform approach). Quarkus already provides a built-in Liveness check and we should probably use it. If Keycloak hangs, the procedure is exactly the same as in WF-based deployments.